### ISSUE TITLE
ONNX export produces incorrect results for variable-length inputs due to data-dependent control flow in LSG attention

### ISSUE THREAD

**Issue by Geeky-Jp007** [ccdv-ai/convert_checkpoint_to_lsg#15]
Description:

LSG XLM-RoBERTa models cannot be reliably exported to ONNX with dynamic sequence lengths. The exported ONNX model produces correct results only for inputs matching the dummy input length used during export, and incorrect results for all other lengths.

Root Cause:

The LSG attention implementation contains data-dependent control flow that gets baked as constants during torch.onnx.export tracing:

modeling_lsg_xlm_roberta.py line 841:
`if t <= 2 * self.block_size + self.num_global_tokens:
`
modeling_lsg_xlm_roberta.py line 973:
`if self.adaptive and t > b and pad > 0:
`
Where t is the runtime sequence length. The ONNX tracer evaluates these conditions once with the dummy input and records only the taken branch. The other branch is completely absent from the exported graph.

Reproduction:

```
from optimum.onnxruntime import ORTModelForSequenceClassification
from transformers import AutoTokenizer
import numpy as np

# Export with short dummy (default ~128 tokens)

model = ORTModelForSequenceClassification.from_pretrained(
    "path/to/lsg-xlm-roberta", export=True, trust_remote_code=True)
model.save_pretrained("onnx_output")
```

# Warning during export

```
/usr/local/lib/python3.12/dist-packages/transformers/utils/generic.py:441: FutureWarning: `torch.utils._pytree._register_pytree_node` is deprecated. Please use `torch.utils._pytree.register_pytree_node` instead.
  _torch_pytree._register_pytree_node(
/usr/local/lib/python3.12/dist-packages/transformers/utils/generic.py:309: FutureWarning: `torch.utils._pytree._register_pytree_node` is deprecated. Please use `torch.utils._pytree.register_pytree_node` instead.
  _torch_pytree._register_pytree_node(
/root/.cache/huggingface/modules/transformers_modules/raw/modeling_lsg_xlm_roberta.py:973: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if self.adaptive and t > b and pad > 0:
/root/.cache/huggingface/modules/transformers_modules/raw/modeling_lsg_xlm_roberta.py:841: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if t <= 2 * self.block_size + self.num_global_tokens:
/root/.cache/huggingface/modules/transformers_modules/raw/modeling_lsg_xlm_roberta.py:149: TracerWarning: Converting a tensor to a Python float might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  attention_scores = query_layer @ key_layer.transpose(-1, -2) / math.sqrt(d)
/usr/local/lib/python3.12/dist-packages/torch/onnx/_internal/torchscript_exporter/jit_utils.py:305: UserWarning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied. (Triggered internally at /pytorch/torch/csrc/jit/passes/onnx/constant_fold.cpp:178.)
  _C._jit_pass_onnx_node_shape_type_inference(node, params_dict, opset_version)
/usr/local/lib/python3.12/dist-packages/torch/onnx/_internal/torchscript_exporter/utils.py:714: UserWarning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied. (Triggered internally at /pytorch/torch/csrc/jit/passes/onnx/constant_fold.cpp:178.)
  _C._jit_pass_onnx_graph_shape_type_inference(
/usr/local/lib/python3.12/dist-packages/torch/onnx/_internal/torchscript_exporter/utils.py:1186: UserWarning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied. (Triggered internally at /pytorch/torch/csrc/jit/passes/onnx/constant_fold.cpp:178.)
  _C._jit_pass_onnx_graph_shape_type_inference(
ONNX export done

```

The ONNX tracer captured the sequence length t as a fixed constant during export. So the exported ONNX model has hardcoded branch decisions based on whatever sequence length the dummy input had during export. For short inputs that happen to match the traced path, it works. For longer inputs that would take a different branch, it produces garbage.

# Load and test
```
import onnxruntime as ort
sess = ort.InferenceSession("onnx_output/model.onnx")
tokenizer = AutoTokenizer.from_pretrained("path/to/lsg-xlm-roberta")
```

# Short input — CORRECT (matches traced path)
```
short = tokenizer("hello", return_tensors="np", max_length=4096, truncation=True)
short_out = sess.run(None, dict(short))  # correct score
```

# Long input — INCORRECT (different branch, not in traced graph)
```
long = tokenizer("long text " * 200, return_tensors="np", max_length=4096, truncation=True)
long_out = sess.run(None, dict(long))  # wrong score
What we tried:
```

Higher ONNX opset versions (14, 17, 21) — same issue, opset doesn't affect tracing behavior
torch.onnx.export with dynamo=True — hangs during export
torch.compile(mode="reduce-overhead") — CUDA graph capture deadlocks for sequences >512 tokens (long LSG path)
torch.compile(mode="max-autotune-no-cudagraphs") — also hangs on long path
Fixed-shape export (no_dynamic_axes=True, sequence_length=4096) — works but forces padding all inputs to 4096, negating LSG's efficiency advantage

ONNX export should support dynamic sequence lengths, or the LSG attention code should provide an export-friendly mode that replaces data-dependent branching with equivalent static operations.

Environment:

PyTorch 2.6+
transformers 4.36.2
optimum (latest)
NVIDIA L4 GPU (24GB)
Triton Inference Server 25.04
Workaround:

Currently using PyTorch + DeepSpeed FP16 inference at ~18ms latency. Fixed-shape ONNX at 4096 works but at ~40ms (worse due to unnecessary padding compute).
